### PR TITLE
Add `kubernetes_manifest` resource

### PR DESCRIPTION
**Pull request by jrhouston**
### Description

This PR adds the `kubernetes_manifest` resource from the [alpha provider](https://github.com/hashicorp/terraform-provider-kubernetes-alpha) to this provider using [terraform-plugin-mux](https://github.com/terraform-plugin-mux). 

The source for the alpha provider is being moved into a subdirectory of this repository so this is a very large PR. 

### Acceptance tests
- [ ] Have you added an acceptance test for the functionality being added?
- [ ] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/CHANGELOG.md):

```release-note
Add kubernetes_manifest resource
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

File: manifest/provider/provider_config.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+				Computed:        false,
+				Sensitive:       false,
+				DescriptionKind: 0,
+				Deprecated:      false,
+			},
+		},
+		BlockTypes: []*tfprotov5.SchemaNestedBlock{
+			{
+				TypeName: "exec",
+				Nesting:  tfprotov5.SchemaNestedBlockNestingModeList,
```
Comment (jrhouston) [inline manifest/provider/provider_config.go:160]: Change this to `SchemaNestedBlockNestingModeSingle`

**Comment by alexsomesan** [comment]
Had a look around and played with it. Works pretty nicely :)

One thing stood out: it crashes when no credentials are specified at all and at least one `kubernetes_manifest` resource is present. The cause seems to be the changes in the "defer client initialization" commit (290ee64d56184f313ffc561445bafabab9d7cfa2). Because it allows an empty client config to be passed to PlanResourceChange, it crashes there. In practice, I don't think we need to do this deferral since the manifest resource cannot do planning without credentials, so there isn't any "further" point to defer towards. I tested with reverting commit: 290ee64d56184f313ffc561445bafabab9d7cfa2 and it works well without it.

This is the configuration I tested with:
```
resource "kubernetes_namespace" "test-ns" {
  metadata {
    name = "test-ns"
  }  
}

resource "kubernetes_manifest" "test-configmap" {
  manifest = {
    "apiVersion" = "v1"
    "kind"       = "ConfigMap"
    "metadata" = {
      "name"      = "test-config"
      "namespace" = kubernetes_namespace.test-ns.metadata.0.name
    }
    "data" = {
      "foo" = "bar"
    }
  }
}
```

The other observation would be around adding `id` to the schema and state. I'm not sure it's worth adding it just for the sake of porting the tests to a testing framework that's being sunset. Of course, if the upcoming testing framework also requires it, then never mind my observation.

**Comment by dak1n1** [comment]
I'm just starting to test this out... It's going to need the changes from [this PR](https://github.com/hashicorp/terraform-provider-kubernetes-alpha/pull/239) to fix this crash:

```
panic: lists must only contain one type of element, saw tftypes.Object["appProtocol":tftypes.String, "name":tftypes.String, "nodePort":tftypes.Number, "port":tftypes.Number, "protocol":tftypes.String, "targetPort":tftypes.Number] and tftypes.Object["appProtocol":tftypes.String, "name":tftypes.String, "nodePort":tftypes.Number, "port":tftypes.Number, "protocol":tftypes.String, "targetPort":tftypes.String]
```

That could be a separate PR, if you prefer. There's still no fix for the underlying IntOrString issue with TargetPort, but at least we can avoid a panic.

**Review by dak1n1 — COMMENTED** [review]
This is a continuation of my previous test, where I was just installing the manifests that are included in the Consul helm chart. I got a couple errors of "inconsistent result after apply", presumably because of the empty lists in the manifests for Roles.

Instead of defining empty lists in the manifests, I had to omit the lists altogether in order for it to apply successfully.

I wonder if there's some way we can smooth out this user experience, because Kubernetes manifests will have empty lists and users will be confused by this.

```
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to
│ kubernetes_manifest.consul["consul/generated/tf-k8s-test-consul-client-role.yaml"],
│ provider "provider[\"localhost/test/kubernetes\"]" produced an unexpected new
│ value: .object.rules: was
│ cty.ListValEmpty(cty.Object(map[string]cty.Type{"apiGroups":cty.List(cty.String),
│ "nonResourceURLs":cty.List(cty.String), "resourceNames":cty.List(cty.String),
│ "resources":cty.List(cty.String), "verbs":cty.List(cty.String)})), but now null.
│
│ This is a bug in the provider, which should be reported in the provider's own
│ issue tracker.
╵
```

Here is the file that caused the error:

```
$ cat  consul/generated/tf-k8s-test-consul-client-role.yaml
# Source: consul/templates/client-role.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: tf-k8s-test-consul-client
  namespace: consul
  labels:
    app: consul
    chart: consul-helm
    heritage: Helm
    release: tf-k8s-test
rules: []
```

**Comment by jrhouston** [comment]
> One thing stood out: it crashes when no credentials are specified at all and at least one kubernetes_manifest resource is present. The cause seems to be the changes in the "defer client initialization" commit (290ee64). Because it allows an empty client config to be passed to PlanResourceChange, it crashes there. In practice, I don't think we need to do this deferral since the manifest resource cannot do planning without credentials, so there isn't any "further" point to defer towards. I tested with reverting commit: 290ee64 and it works well without it.

Makes sense – I'm going to revert that commit.

> The other observation would be around adding id to the schema and state. I'm not sure it's worth adding it just for the sake of porting the tests to a testing framework that's being sunset. Of course, if the upcoming testing framework also requires it, then never mind my observation.

It's not clear to me from looking at the new framework repo if resources are going to require an ID. I'm going to revert and we can continue to test by running `make testacc` inside the manifest directory.

**Comment by jrhouston** [comment]
I've added the code to put the manifest resource behind an experiment flag. Here's what the user will see if they try to use it when it's not enabled. 

```
╷
│ Error: Experimental feature not enabled.
│
│   with kubernetes_manifest.example,
│   on main.tf line 15, in resource "kubernetes_manifest" "example":
│   15: resource kubernetes_manifest example {
│
│ The `kubernetes_manifest` resource is an experimental feature and must be explicitly enabled in the provider configuration block.
╵
```

File: go.mod [hunk 2]
```diff
@@ -19,38 +19,46 @@ require (
...
+	github.com/mitchellh/hashstructure v1.1.0
 	github.com/mitchellh/mapstructure v1.4.1 // indirect
-	github.com/oklog/run v1.1.0 // indirect
 	github.com/robfig/cron v1.2.0
 	github.com/russross/blackfriday v1.6.0 // indirect
 	github.com/stretchr/objx v0.3.0 // indirect
+	github.com/stretchr/testify v1.7.0
 	github.com/ulikunitz/xz v0.5.10 // indirect
 	github.com/xlab/treeprint v1.1.0 // indirect
+	github.com/zclconf/go-cty v1.8.4
```
[the remark this answers is not in the document]
Comment (jrhouston) [reply go.mod:48]: It's a dependency of the hcl package which we use to parse the field paths for the waiter, so we can't just swap it out.

File: .github/workflows/manifest_unit.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    branches:
+      - master
+    paths:
+      - "**.go"
+jobs:
+  unit_tests:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (jrhouston) [reply .github/workflows/manifest_unit.yaml:19]: This document is talking about 3rd party actions, and says: 

> Pin actions to a tag only if you trust the creator

Given GitHub is the maintainer of the checkout action I think we should be OK.

**Comment by alexsomesan** [comment]
> > One thing stood out: it crashes when no credentials are specified at all and at least one kubernetes_manifest resource is present. The cause seems to be the changes in the "defer client initialization" commit ([290ee64](https://github.com/hashicorp/terraform-provider-kubernetes/commit/290ee64d56184f313ffc561445bafabab9d7cfa2)). Because it allows an empty client config to be passed to PlanResourceChange, it crashes there. In practice, I don't think we need to do this deferral since the manifest resource cannot do planning without credentials, so there isn't any "further" point to defer towards. I tested with reverting commit: [290ee64](https://github.com/hashicorp/terraform-provider-kubernetes/commit/290ee64d56184f313ffc561445bafabab9d7cfa2) and it works well without it.
> 
> Makes sense – I'm going to revert that commit.

Thanks!

> 
> > The other observation would be around adding id to the schema and state. I'm not sure it's worth adding it just for the sake of porting the tests to a testing framework that's being sunset. Of course, if the upcoming testing framework also requires it, then never mind my observation.
> 
> It's not clear to me from looking at the new framework repo if resources are going to require an ID. I'm going to revert and we can continue to test by running `make testacc` inside the manifest directory for now.

I've talked to the SDK team and it's not yet known to them either whether an ID will be used or not.  Let's hold on adding ID until they have made their design decision.

File: go.mod [hunk 4]
```diff
@@ -5,10 +5,10 @@ require (
 	github.com/Azure/go-autorest/autorest v0.11.18 // indirect
 	github.com/MakeNowJust/heredoc v1.0.0 // indirect
 	github.com/aws/aws-sdk-go v1.38.20 // indirect
+	github.com/davecgh/go-spew v1.1.1
```
[the remark this answers is not in the document]
Comment (alexsomesan) [reply go.mod:8]: We can remove go-spew altogether. I used it for dumping complex data structures to the log during development, but I think we can do without it.

**Comment by alexsomesan** [comment]
> This is a continuation of my previous test, where I was just installing the manifests that are included in the Consul helm chart. I got a couple errors of "inconsistent result after apply", presumably because of the empty lists in the manifests for Roles.
> 
> Instead of defining empty lists in the manifests, I had to omit the lists altogether in order for it to apply successfully.
> 
> I wonder if there's some way we can smooth out this user experience, because Kubernetes manifests will have empty lists and users will be confused by this.
> 
> ```
> │ Error: Provider produced inconsistent result after apply
> │
> │ When applying changes to
> │ kubernetes_manifest.consul["consul/generated/tf-k8s-test-consul-client-role.yaml"],
> │ provider "provider[\"localhost/test/kubernetes\"]" produced an unexpected new
> │ value: .object.rules: was
> │ cty.ListValEmpty(cty.Object(map[string]cty.Type{"apiGroups":cty.List(cty.String),
> │ "nonResourceURLs":cty.List(cty.String), "resourceNames":cty.List(cty.String),
> │ "resources":cty.List(cty.String), "verbs":cty.List(cty.String)})), but now null.
> │
> │ This is a bug in the provider, which should be reported in the provider's own
> │ issue tracker.
> ╵
> ```
> 
> Here is the file that caused the error:
> 
> ```
> $ cat  consul/generated/tf-k8s-test-consul-client-role.yaml
> # Source: consul/templates/client-role.yaml
> apiVersion: rbac.authorization.k8s.io/v1
> kind: Role
> metadata:
>   name: tf-k8s-test-consul-client
>   namespace: consul
>   labels:
>     app: consul
>     chart: consul-helm
>     heritage: Helm
>     release: tf-k8s-test
> rules: []
> ```

Thanks for bringing this up. I'll look for a solution here. I can't remember off the cuff if there were any resources where empty lists are of special significance. Otherwise we should be filtering them out.

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrhouston) [reply go.mod:8]: Yeah – I'll remove the go-spew stuff it was just debug logs.

[the remark this answers is not in the document]
Comment (jrhouston) [reply go.mod]: I've changed it to just dump the `%#v` value using hcllog.Fmt.

**Pull request merged**